### DISCUSSION TITLE
Running with Launchd - avoiding permission problems

### DISCUSSION THREAD

**Discussion by synox** [RhetTbull/osxphotos#301] [Show and tell]
Hi all

osxphotos works great when called directly from the terminal. Initially I had some trouble running it with launchd, but wrapping it into a automator application solved the permissions problems. Here I share how I made it work. 

Bash-script: /Users/johndoe/bin/export-photos.bash
```bash
#!/bin/bash

# pip3 install osxphotos
 
/usr/local/bin/python3 -m osxphotos export \
	--update  --use-photokit   --touch-file  --deleted  --download-missing  \
	--directory "{created.date} {place.name.city,} {place.country_code,}" \
		"/Users/johndoe/Pictures/Photos Library.photoslibrary" "/Users/johndoe/Pictures/Apple-Photos-Exported"
```

Automator-App: /Users/johndoe/bin/export-photos.app
"Run Shell Script": 
```bash
/Users/johndoe/bin/export-photos.bash >> /Users/johndoe/bin/export-photos.bash-out.log 2>> /Users/johndoe/bin/export-photos.bash-err.log
```

Launchd config: /Users/johndoe/bin/export-photos.plist
```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
  <key>Label</key>
  <string>aravindo.export-photos</string>

  <key>ProgramArguments</key>
  <array>
    <string>open</string>
    <string>-W</string>
    <string>--background</string>
    <string>-a</string>
    <string>/Users/aravindo/bin/export-photos.app</string>
  </array>

 
  <key>RunAtLoad</key>
  <true/>
 
  <key>StartInterval</key>
  <integer>3600</integer>
  
  <key>StandardErrorPath</key>
  <string>/Users/aravindo/bin/export-photos.bash-err.log</string>

  <key>StandardOutPath</key>
  <string>/Users/aravindo/bin/export-photos.bash-out.log</string>
</dict>
</plist>
```

Then to install the launchd config: 
```
mkdir /Users/johndoe/Pictures/Apple-Photos-Exported
cd $HOME/Library/LaunchAgents  
ln -s /Users/johndoe/bin/export-photos.plist export-photos.plist
launchctl unload $HOME/Library/LaunchAgents/export-photos.plist
launchctl load $HOME/Library/LaunchAgents/export-photos.plist
```

I added `export-photos.app` to "Full Disk Access", just to be sure. 
 
Hope that works for you, merry xmas!

**Comment by RhetTbull** [comment]
Thanks @synox for sharing!  Others have had problems with permissions especially when using --use-photokit as the calling process must first be authorized to access Photos.  This looks like a good solution!  I'm guessing Automator is pre-authorized to access Photos.

**Reply by synox** [reply]
I think you can first run the automator generated app manually, and then it will show a permission dialog. Then it should work for launchd. I tried so many things that I am not sure any more.
